pyOxygenSCPI/oxygenscpi.py

```diff
@@ -532,31 +579,162 @@ def setElogPeriod(self, period):
-            return self._sendRaw(':ELOG:TIM ABS')
-        return self._sendRaw(':ELOG:TIM OFF')
+    @contextmanager
+    def elogContext(self):
+        """Safely starts and stops external logging.
+
+        This function should be used in a with statement to start external
+        logging and immediately stops it when either exiting the context
+        or when an Exception occurs within the context.
+
+        Example usage:
+            with mDevice.startElog():
+                # Here elog is started
+                time.sleep(10)
+                data = mDevice.fetchElog()
+            # Here elog is stopped
+        """
+        try:
+            self.startElog()
+            yield
+        finally:
+            self.stopElog()
 
-    def fetchElog(self):
+    def _getElogTimestamp(self):
+        """Get external logging configured timestamp.
+
+        Returns:
+            External logging timestamp string obtained from ':ELOG:TIM?'
+        """
+        ret = self._askRaw(':ELOG:TIM?')
+        if isinstance(ret, bytes):
+            ret = ret.decode().strip()
+            self.elogTimestamp = ret
+            return ret
+        return False
+
+    def setElogTimestamp(self, tsType='REL'):
+        if tsType in ('REL', 'ABS', 'ELOG'):
+            self._sendRaw(f':ELOG:TIM {tsType}')
+            ts_read = self._getElogTimestamp()
+            return ts_read == tsType
+        send_ok = self._sendRaw(':ELOG:TIM OFF')
+        ts_read = self._getElogTimestamp()
+        return send_ok
+
+    def fetchElog(self, raw_string=True):
         data = self._askRaw(':ELOG:FETCH?')
         if type(data) is bytes:
             data = data.decode()
         else:
             return False
-        if 'NONE' in data:
+        if any(d in data for d in ('NONE', 'ERROR')):
             return False
         # Remove Header if Whitespace present
         if ' ' in data:
             data = data.split(' ')[1]
         data = data.split(',')
-        num_ch = len(self.elogChannelList)+1
+        num_ch = len(self.elogChannelList)
+        if self.elogTimestamp in ('REL', 'ABS', 'ELOG'):
+            num_ch += 1
         #print(len(data)/(1.0*num_ch), data)
         num_items = int(len(data)/num_ch)
         data = [data[i*num_ch:i*num_ch+num_ch] for i in range(num_items)]
+        if not raw_string:
+            for i, row in enumerate(data):
+                data[i] = self._convertElogArray(row)
         return data
 
+    def _convertElogArray(self, data_array):
+        """Converts a single array from fetchElog string values into float.
+        
+        If the Elog timestamp is set to 'ABS' then the first value of the array
+        is converted into datetime object.
+
+        Args:
+        data_array : list of strings
+            List containing single array of string measurements from fetchElog.
+
+        Returns:
+            List with value types converted to float (datetime for value at
+            index 0 if timestamp is ABS).
```

Comment:
Ideally, the return type should be something like a pandas DataFrame where columns can be labelled. Using lists is probably fine for now but DataFrames are more convenient for new users and it"s clear which column contains what data.
If you need array/list data to work, we could also add an extensible "output-format" parameter to fetchElogAccumulated and fetchElog (instead of raw_strings) which could be "raw", "list", "DataFrame" and so on.

Reply:
Yes definitely, I was using pandas and numpy for the same functionality in previous projects. The "output-format" is a very good option.
One thing to notice is that those are external packages, so they should be in the setuptools `install_requires` to be installed altogether with the pyOxygenSCPI